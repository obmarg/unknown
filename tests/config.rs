// TODO: At some point use libtest-mimic to make this smarter?  CBA now.
use assert_cmd::Command;

#[test]
fn task_require_non_ancestor() {
    test_failing_config("task_require_non_ancestor");
}

#[test]
fn task_require_unknown_project() {
    test_failing_config("task_require_unknown_project");
}

fn test_failing_config(name: &str) {
    let mut cmd = Command::cargo_bin("unknown").unwrap();
    cmd.arg("projects");
    cmd.current_dir(format!("tests/config-test-cases/{name}"));

    let assert = cmd.assert().failure();
    let stdout = String::from_utf8_lossy(&assert.get_output().stdout);
    let stderr = String::from_utf8_lossy(&assert.get_output().stderr);

    insta::assert_snapshot!(format!("{name}.stdout"), stdout.as_ref());
    insta::assert_snapshot!(format!("{name}.stderr"), stderr.as_ref());
}
