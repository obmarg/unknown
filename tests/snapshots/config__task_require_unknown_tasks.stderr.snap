---
source: tests/config.rs
expression: stderr.as_ref()
---
Error: 
  × Errors occurred when validating your configuration

Error: 
  × Found a requires statement that doesn't match any tasks
    ╭─[project.kdl:8:1]
  8 │     task "build" {
  9 │         requires "pre-build"
    ·                  ─────┬─────
    ·                       ╰── expected to find at least one task with this name in service-a
 10 │         requires "pre-build-2" in="self"
    ╰────
  help: Make sure you've specified the correct task name
Error: 
  × Found a requires statement that doesn't match any tasks
    ╭─[project.kdl:9:1]
  9 │         requires "pre-build"
 10 │         requires "pre-build-2" in="self"
    ·                  ──────┬──────    ───┬──
    ·                        │             ╰── this project does not have a task named pre-build-2
    ·                        ╰── expected to find at least one task with this name
 11 │         requires "pre-build-3" in="^self"
    ╰────
  help: Make sure you've specified the correct project and task name
Error: 
  × Found a requires statement that doesn't match any tasks
    ╭─[project.kdl:10:1]
 10 │         requires "pre-build-2" in="self"
 11 │         requires "pre-build-3" in="^self"
    ·                  ──────┬──────    ───┬───
    ·                        │             ╰── these projects do not have a task named pre-build-3
    ·                        ╰── expected to find at least one task with this name
 12 │         requires "pre-build-4" in="other-project"
    ╰────
  help: Make sure you've specified the correct project and task name
Error: 
  × Found a requires statement that doesn't match any tasks
    ╭─[project.kdl:11:1]
 11 │         requires "pre-build-3" in="^self"
 12 │         requires "pre-build-4" in="other-project"
    ·                  ──────┬──────    ───────┬───────
    ·                        │                 ╰── this project does not have a task named pre-build-4
    ·                        ╰── expected to find at least one task with this name
 13 │     }
    ╰────
  help: Make sure you've specified the correct project and task name


